integration-tests/src/main/java/org/apache/druid/testing/clients/AbstractQueryResourceTestClient.java

```diff
@@ -21,67 +21,157 @@
 
 import com.fasterxml.jackson.core.type.TypeReference;
 import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.jaxrs.smile.SmileMediaTypes;
 import com.google.inject.Inject;
+import org.apache.druid.guice.annotations.Smile;
+import org.apache.druid.java.util.common.IAE;
 import org.apache.druid.java.util.common.ISE;
 import org.apache.druid.java.util.http.client.HttpClient;
 import org.apache.druid.java.util.http.client.Request;
+import org.apache.druid.java.util.http.client.response.BytesFullResponseHandler;
+import org.apache.druid.java.util.http.client.response.BytesFullResponseHolder;
 import org.apache.druid.java.util.http.client.response.StatusResponseHandler;
 import org.apache.druid.java.util.http.client.response.StatusResponseHolder;
-import org.apache.druid.testing.IntegrationTestingConfig;
 import org.apache.druid.testing.guice.TestClient;
+import org.jboss.netty.handler.codec.http.HttpHeaders;
 import org.jboss.netty.handler.codec.http.HttpMethod;
 import org.jboss.netty.handler.codec.http.HttpResponseStatus;
 
+import javax.ws.rs.core.MediaType;
+import java.io.IOException;
 import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.Future;
 
 public abstract class AbstractQueryResourceTestClient<QueryType>
 {
-  private final ObjectMapper jsonMapper;
-  private final HttpClient httpClient;
+  private String contentTypeHeader = MediaType.APPLICATION_JSON;
+
+  /**
+   * a 'null' means the Content-Type in response defaults to Content-Type of request
+   */
+  private String acceptHeader = null;
+
+  final ObjectMapper jsonMapper;
+  final ObjectMapper smileMapper;
+  final HttpClient httpClient;
   final String routerUrl;
+  final Map<String, EncoderDecoder<QueryType>> encoderDecoderMap;
```

Comment:
nit: out of curiosity, are this map and the fancy `EncoderDecoder` interface for more extendable structure where we want to support new formats in the future?

Reply:
It's possible since the generic type 'QueryType' is defined on outer class and used by `query` function which is the caller of this `EncoderDecoder`. 

This generic contract has been removed in the new commit, because I found that its implementation has nothing to do with type info.